pkg/csi/csi.go

```diff
@@ -58,19 +59,35 @@ func InitializeSidecars(ver string) error {
 		return err
 	}
 
-	sv := version.MustParseGeneric(ver)
-
-	for _, image := range si {
-		v := version.MustParseGeneric(image.Version)
-		if sv.Major() == v.Major() && sv.Minor() == v.Minor() {
-			sidecarImages = &image
-			break
+	pltVersion := version.MustParseGeneric(ver)
+
+	// this helps in not parsing the version twice in loop.
+	closestVersion := pltVersion.WithMinor(0)
+	for idx := range si {
+		siVersion := version.MustParseGeneric(si[idx].Version)
+		log.Info("searching for the most compatible CSI image version", "CSI", siVersion, "Platform", pltVersion)
+
+		// only check sidecar image versions that are not higher than platform
+		if siVersion.Major() == pltVersion.Major() && siVersion.Minor() <= pltVersion.Minor() {
+			// filter sidecar closest to platform version
+			if siVersion.Minor() >= closestVersion.Minor() {
+				sidecarImages = &si[idx]
+				closestVersion = siVersion
+			}
+			if closestVersion.Minor() == pltVersion.Minor() && sidecarImages != nil { // exact match and early exit
+				break
+			}
+		} else {
+			log.Info("skipping sidecar images: version greater than platform version")
```

Comment:
The code only uses the closest version for the minor part and never considers the rest of its parts. 
As such, it will be easier to treat it as an int and use `-1` to generalize the edge case.

```suggestion
	closestMinor := -1
	for idx := range si {
		siVersion := version.MustParseGeneric(si[idx].Version)
		log.Info("searching for the most compatible CSI image version", "CSI", siVersion, "Platform", pltVersion)

		// only check sidecar image versions that are not higher than platform
		if siVersion.Major() == pltVersion.Major() && siVersion.Minor() <= pltVersion.Minor() {
			// filter sidecar closest to platform version
			if siVersion.Minor() > closestMinor {
				sidecarImages = &si[idx]
				closestMinor = siVersion.Minor()
			}
			if closestMinor == pltVersion.Minor {
				break
			}
		} else {
			log.Info("skipping sidecar images: version greater than platform version")
```

Reply:
done.